File: android/src/main/java/com/reactnativekalturaplayer/KalturaPlayerModule.kt
package com.reactnativekalturaplayer

import android.text.TextUtils
import androidx.annotation.NonNull
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.kaltura.playkit.PKLog
import com.kaltura.tvplayer.KalturaPlayer

class KalturaPlayerModule(
    reactApplicationContext: ReactApplicationContext,
    kalturaPlayerViewManager: KalturaPlayerViewManager?) : ReactContextBaseJavaModule(reactApplicationContext) {

    private val log = PKLog.get(KalturaPlayerModule::class.java.simpleName)
    private val PLAYER_CLASS = "KalturaPlayerModule"
    private lateinit var kalturaPlayerRNView: KalturaPlayerRNView

    private var context: ReactApplicationContext = reactApplicationContext
    private val kalturaPlayerRN: KalturaPlayerRN

    init {
        if (kalturaPlayerViewManager != null) {
            kalturaPlayerRNView = kalturaPlayerViewManager.kalturaPlayerView
        }

        kalturaPlayerRN = KalturaPlayerRN(context, kalturaPlayerRNView)
    }

    override fun getName(): String {
        return PLAYER_CLASS
    }

    @ReactMethod
    fun setUpPlayer(playerType: String, partnerId: Int = 0, initOptions: String?, promise: Promise) {
        log.d("setPartnerId: $partnerId")
        checkArguments(this::setUpPlayer.name, playerType)
        kalturaPlayerRN.createPlayerInstance(getKalturaPlayerType(playerType), partnerId, initOptions, promise)
    }

    @ReactMethod
    fun load(assetId: String?, mediaAsset: String?, promise: Promise) {
        log.d(
            "load assetId: " + assetId +
                    "\n , mediaAssetJson:" + mediaAsset
        )
        checkArguments(this::load.name, assetId)
        kalturaPlayerRN.load(assetId, mediaAsset, promise)
    }

    @ReactMethod
    fun addPlayerView() {
        log.d("addPlayerView")
        kalturaPlayerRN.addPlayerView()
    }

    @ReactMethod
    fun removePlayerView() {
        log.d("removePlayerView")
        kalturaPlayerRN.removePlayerView()
    }

    @ReactMethod
    fun onApplicationPaused() {
        log.d("onApplicationPaused")
        kalturaPlayerRN.onApplicationPaused()
    }

    @ReactMethod
    fun onApplicationResumed() {
        log.d("onApplicationResumed")
        kalturaPlayerRN.onApplicationResumed()
    }

    @ReactMethod
    fun updatePluginConfigs(pluginConfigJson: String?) {
        log.d("updatePluginConfigs")
        checkArguments(this::updatePluginConfigs.name, pluginConfigJson)
        kalturaPlayerRN.updatePluginConfigs(pluginConfigJson)
    }

    @ReactMethod
    fun play() {
        log.d("play")
        kalturaPlayerRN.play()
    }

    @ReactMethod
    fun pause() {
        log.d("pause")
        kalturaPlayerRN.pause()
    }

    @ReactMethod
    fun replay() {
        log.d("replay")
        kalturaPlayerRN.replay()
    }

    @ReactMethod
    fun seekTo(position: Float) {
        log.d("seekTo:$position")
        if (position > 0f) {
            kalturaPlayerRN.seekTo(position);
        } else {
            log.d("Invalid seek position which is $position")
        }
    }

    @ReactMethod
    fun changeTrack(uniqueId: String?) {
        log.d("changeTrack:$uniqueId")
        checkArguments(this::changeTrack.name, uniqueId)
        kalturaPlayerRN.changeTrack(uniqueId)
    }

    @ReactMethod
    fun changePlaybackRate(playbackRate: Float) {
        log.d("changePlaybackRate:$playbackRate")
        if (playbackRate > 0f) {
            kalturaPlayerRN.changePlaybackRate(playbackRate);
        } else {
            log.d("Invalid playback rate which is $playbackRate");
        }
    }

    @ReactMethod
    fun destroy() {
        log.d("destroy")
        kalturaPlayerRN.destroy()
    }

    @ReactMethod
    fun stop() {
        log.d("stop")
        kalturaPlayerRN.stop()
    }

    @ReactMethod
    fun setAutoplay(autoplay: Boolean) {
        log.d("setAutoplay: $autoplay")
        kalturaPlayerRN.setAutoplay(autoplay)
    }

    @ReactMethod
    fun setKS(ks: String) {
        log.d("setKS: $ks")
        checkArguments(this::setKS.name, ks)
        kalturaPlayerRN.setKS(ks)
    }

    //TODO: If required expose it
    fun setZIndex(index: Float) {
        log.d("setZIndex: $index")
        kalturaPlayerRN.setZIndex(index)
    }

    @ReactMethod
    fun setVolume(volume: Float) {
        log.d("setVolume: $volume")
        if (volume >= 0f) {
            kalturaPlayerRN.setVolume(volume);
        } else {
            log.d("Invalid Volume which is $volume");
        }
    }

    @ReactMethod
    fun seekToLiveDefaultPosition() {
        log.d("seekToLiveDefaultPosition")
        kalturaPlayerRN.seekToLiveDefaultPosition()
    }

    @ReactMethod
    fun updateSubtitleStyle(subtitleStyleSettings: String?) {
        log.d("updateSubtitleStyle")
        checkArguments(this::updateSubtitleStyle.name, subtitleStyleSettings)
        kalturaPlayerRN.updateSubtitleStyle(subtitleStyleSettings)
    }

    @ReactMethod
    fun updateResizeMode(resizeMode: String) {
        log.d("updateResizeMode")
        checkArguments(this::updateResizeMode.name, resizeMode)
        kalturaPlayerRN.updateResizeMode(resizeMode)
    }

    @ReactMethod
    fun updateAbrSettings(abrSettings: String?) {
        log.d("updateAbrSettings")
        checkArguments(this::updateAbrSettings.name, abrSettings)
        kalturaPlayerRN.updateAbrSettings(abrSettings)
    }

    @ReactMethod
    fun resetAbrSettings() {
        log.d("resetAbrSettings")
        kalturaPlayerRN.resetAbrSettings()
    }

    @ReactMethod
    fun updateLLConfig(pkLowLatencyConfig: String?) {
        log.d("updateLLConfig")
        checkArguments(this::updateLLConfig.name, pkLowLatencyConfig)
        kalturaPlayerRN.updateLLConfig(pkLowLatencyConfig)
    }

    @ReactMethod
    fun resetLLConfig() {
        log.d("resetLLConfig")
        kalturaPlayerRN.resetLLConfig()
    }

    @ReactMethod
    fun getCurrentPosition(promise: Promise) {
        log.d("getCurrentPosition")
        kalturaPlayerRN.getCurrentPosition(promise)
    }

    @ReactMethod
    fun isPlaying(promise: Promise) {
        log.d("isPlaying")
        kalturaPlayerRN.isPlaying(promise)
    }

    @ReactMethod
    fun isLive(promise: Promise) {
        log.d("isLive")
        kalturaPlayerRN.isLive(promise)
    }

    @ReactMethod
    fun removeKalturaPlayerListeners() {
        log.d("removeKalturaPlayerListeners")
        kalturaPlayerRN.removeKalturaPlayerListeners()
    }

    @ReactMethod
    fun addKalturaPlayerListeners() {
        log.d("addKalturaPlayerListeners")
        kalturaPlayerRN.addKalturaPlayerListeners()
    }

    @ReactMethod
    fun requestThumbnailInfo(positionMs: Float/** position is milliseconds **/, promise: Promise) {
        log.d("requestThumbnailInfo position is $positionMs")
        kalturaPlayerRN.requestThumbnailInfo(positionMs, promise)
    }

    @ReactMethod
    fun setLogLevel(logLevel: String) {
        log.d("setLogLevel $logLevel")
        kalturaPlayerRN.setLogLevel(logLevel)
    }

    @Throws(IllegalArgumentException::class)
    @NonNull
    private fun checkArguments(methodName: String, arg: String?) {
        if (TextUtils.isEmpty(arg)) {
            throw IllegalArgumentException("$methodName argument is invalid for ", Throwable("$arg is invalid"))
        }
    }

    @NonNull
    private fun getKalturaPlayerType(playerType: String): KalturaPlayer.Type {
        if (TextUtils.equals(playerType, KalturaPlayer.Type.basic.name)) {
            return KalturaPlayer.Type.basic
        } else if (TextUtils.equals(playerType, KalturaPlayer.Type.ovp.name)) {
            return KalturaPlayer.Type.ovp
        }
        return KalturaPlayer.Type.ott
    }
}

File: ios/KalturaPlayerModule.swift
//
//  KalturaPlayerModule.swift
//  RNKalturaPlayer
//
//  Created by Nilit Danan on 15/06/2022.
//

import Foundation
import PlayKit

@objc(KalturaPlayerEvents)
class KalturaPlayerEvents: RCTEventEmitter {
    public static var emitter: RCTEventEmitter!
    var hasListener: Bool = false

    override init() {
        super.init()
        KalturaPlayerEvents.emitter = self
    }
    
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }

    override func startObserving() {
        hasListener = true
    }

    override func stopObserving() {
        hasListener = false
    }

    override func supportedEvents() -> [String] {
        let playerEvents: [String] = KalturaPlayerRNEvents.allCases.map { $0.rawValue }
        let adEvents: [String] = KalturaPlayerRNAdEvents.allCases.map { $0.rawValue }
        let analyticsEvents: [String] = KalturaPlayerRNAnalyticsEvents.allCases.map { $0.rawValue }
        // If We don't return all supported Events in the lib, and the app listens to it, it will crash!
        let unsupportedPlayerEvents: [String] = KalturaPlayerUnsupportedRNEvents.allCases.map { $0.rawValue }
        let unsupportedAdEvents: [String] = KalturaPlayerUnsupportedRNAdEvents.allCases.map { $0.rawValue }
        
        return playerEvents + adEvents + analyticsEvents + unsupportedPlayerEvents + unsupportedAdEvents
    }
}

@objc(KalturaPlayerModule)
class KalturaPlayerModule: NSObject, RCTBridgeModule {
    
    var bridge: RCTBridge!
    var playerType: PlayerType = .basic
    var kalturaPlayerRN: KalturaPlayerRN?
    
    static func moduleName() -> String! {
        return "KalturaPlayerModule"
    }
    
    // Special method in order to expose static data.
    // Will enable to see the object in the JS console.
    // You can return any kind of data, but keep in mind that this data is static, computed on build.
    // This means that if you change this data on runtime, you won’t get the updated values.
    @objc func constantsToExport() -> [AnyHashable : Any]! {
        return [:]
    }
    
    @objc static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    @objc func setUpPlayer(_ type: String, partnerId: Int = 0, initOptions: String?,
                           resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        
        self.playerType = getPlayerType(from: type)
        var defaultOptions = initOptions
        
        if playerType == .basic {
            if initOptions == nil {
                defaultOptions = "{}"
            } else {
                if let initOptions = initOptions, initOptions.isEmpty {
                    defaultOptions = "{}"
                }
            }
        }
        
        guard let options = defaultOptions, !options.isEmpty else {
            let message = "The initOptions can not be empty."
            let error = KalturaPlayerRNError.setupFailed(message: message).asNSError
            reject("ERROR_SETUPPLAYER", message, error)
            return
        }
        
        let data = Data(options.utf8)
        let initOptions: RNKPInitOptions
        do {
            initOptions = try JSONDecoder().decode(RNKPInitOptions.self, from: data)
        } catch let error as NSError {
            let message = "The initOptions could not be parsed. \(error.localizedDescription)"
            let error = KalturaPlayerRNError.setupFailed(message: message).asNSError
            reject("ERROR_SETUPPLAYER", message, error)
            return
        }
        
        //TODO: Remove print
        print(initOptions)
        
        DispatchQueue.main.async {
            switch self.playerType {
            case .basic:
                self.kalturaPlayerRN = BasicKalturaPlayerRN(withOptions: initOptions)
            case .ott:
                self.kalturaPlayerRN = OTTKalturaPlayerRN(withOptions: initOptions, partnerId: partnerId)
            case .ovp:
                self.kalturaPlayerRN = OVPKalturaPlayerRN(withOptions: initOptions, partnerId: partnerId)
            }
            
            // Connect the player view
            guard self.bridge.moduleIsInitialized(KalturaPlayerViewManager.self),
                    let playerViewManager = self.bridge.module(for: KalturaPlayerViewManager.self) as? KalturaPlayerViewManager else {
                let message = "The KalturaPlayerViewManager was not yet initialised."
                let error = KalturaPlayerRNError.setupFailed(message: message).asNSError
                reject("ERROR_SETUPPLAYER", message, error)
                return
            }
            
            self.kalturaPlayerRN?.connectView(playerViewManager.kalturaPlayerRNView.kalturaPlayerView)
            
            self.kalturaPlayerRN?.updateViewContentMode(initOptions.aspectRatioResizeMode)
            
            resolve("Sucess")
        }
    }
    
    @objc func load(_ assetId: String?, mediaAsset: String?,
                    resolver resolve: @escaping RCTPromiseResolveBlock,
                    rejecter reject: @escaping RCTPromiseRejectBlock) {
        
        guard let kalturaPlayerRN = kalturaPlayerRN else {
            let message = "The KalturaPlayerRN is nil."
            let error = KalturaPlayerRNError.loadMediaFailed(message: message).asNSError
            reject("ERROR_LOADMEDIA", message, error)
            return
        }
        
        guard let assetId = assetId, !assetId.isEmpty else {
            let message = "The assetId is empty."
            let error = KalturaPlayerRNError.loadMediaFailed(message: message).asNSError
            reject("ERROR_LOADMEDIA", message, error)
            return
        }
        
        let playbackURL = URL(string: assetId)
        let isValidURL = UIApplication.shared.canOpenURL(playbackURL!)
        
        // If FE is passing an assetId which is playback URL actually then try to play it using basic player way
        if isValidURL && playerType != .basic {
            DispatchQueue.main.async {
                kalturaPlayerRN.playerMediaUsingBasicPlayer(assetId: assetId, mediaAsset: mediaAsset ?? "{}") { error in
                    if let kpRNError = error {
                        let message = kpRNError.userInfo[KalturaPlayerRNError.errorMessageKey] as? String
                        let nsError = kpRNError.asNSError
                        reject("ERROR_LOADMEDIA", message, nsError)
                    } else {
                        resolve("Sucess")
                    }
                }
            }
            return
        }
        
        DispatchQueue.main.async {
            kalturaPlayerRN.load(assetId: assetId, mediaAsset: mediaAsset ?? "{}") { error in
                if let kpRNError = error {
                    let message = kpRNError.userInfo[KalturaPlayerRNError.errorMessageKey] as? String
                    let nsError = kpRNError.asNSError
                    reject("ERROR_LOADMEDIA", message, nsError)
                } else {
                    resolve("Sucess")
                }
            }
        }
    }
}

extension KalturaPlayerModule {
    
    enum PlayerType: String {
        case basic
        case ovp
        case ott
    }
    
    private func getPlayerType(from type: String) -> PlayerType {
        
        if let playerType = PlayerType(rawValue: type.lowercased()) {
            return playerType
        }
        
        return .basic
    }
}

extension KalturaPlayerModule {
    
    @objc func addKalturaPlayerListeners() {
        kalturaPlayerRN?.observeAllEvents()
    }
    
    @objc func removeKalturaPlayerListeners() {
        kalturaPlayerRN?.removeObservationForAllEvents()
    }
}

extension KalturaPlayerModule {
    
    @objc func onApplicationPaused() {
        // TODO: onApplicationPaused
    }
    
    @objc func onApplicationResumed() {
        // TODO: onApplicationResumed
    }
}

extension KalturaPlayerModule {
    
    @objc func updatePluginConfigs(_ configs: String?) {
        guard let pluginConfigs = configs, !pluginConfigs.isEmpty else { return }
        
        let data = Data(pluginConfigs.utf8)
        if let plugins = try? JSONDecoder().decode(Plugins.self, from: data) {
            kalturaPlayerRN?.updatePlugins(plugins)
        }
    }
}

extension KalturaPlayerModule {
    
    @objc func play() {
        DispatchQueue.main.async { [weak self] in
            self?.kalturaPlayerRN?.kalturaPlayer?.play()
        }
    }
    
    @objc func pause() {
        DispatchQueue.main.async { [weak self] in
            self?.kalturaPlayerRN?.kalturaPlayer?.pause()
        }
    }
    
    @objc func replay() {
        DispatchQueue.main.async { [weak self] in
            self?.kalturaPlayerRN?.kalturaPlayer?.replay()
        }
    }
    
    @objc func seekTo(_ position: TimeInterval) {
        DispatchQueue.main.async { [weak self] in
            self?.kalturaPlayerRN?.kalturaPlayer?.seek(to: position)
        }
    }
    
    @objc func seekToLiveDefaultPosition() {
        DispatchQueue.main.async { [weak self] in
            self?.kalturaPlayerRN?.kalturaPlayer?.seekToLiveEdge()
        }
    }
    
    @objc func stop() {
        DispatchQueue.main.async { [weak self] in
            self?.kalturaPlayerRN?.kalturaPlayer?.stop()
        }
    }
    
    @objc func destroy() {
        DispatchQueue.main.async { [weak self] in
            self?.kalturaPlayerRN?.kalturaPlayer?.destroy()
        }
    }
    
    @objc func changeTrack(_ trackId: String?) {
        guard let trackId = trackId, !trackId.isEmpty else { return }
        DispatchQueue.main.async { [weak self] in
            self?.kalturaPlayerRN?.kalturaPlayer?.selectTrack(trackId: trackId)
        }
    }
    
    @objc func changePlaybackRate(_ playbackRate: Float) {
        DispatchQueue.main.async { [weak self] in
            self?.kalturaPlayerRN?.kalturaPlayer?.rate = playbackRate
        }
    }
    
    @objc func setVolume(_ volume: Float) {
        DispatchQueue.main.async { [weak self] in
            self?.kalturaPlayerRN?.kalturaPlayer?.volume = volume
        }
    }
}

extension KalturaPlayerModule {
    
    @objc func updateResizeMode(_ mode: String?) {
        DispatchQueue.main.async { [weak self] in
            self?.kalturaPlayerRN?.updateViewContentMode(mode)
        }
    }
    
    @objc func updateAbrSettings(_ abrSettings: String?) {
        guard let settings = abrSettings, !settings.isEmpty else { return }
        
        let data = Data(settings.utf8)
        let abrSettings = try? JSONDecoder().decode(ABRSettings.self, from: data)
        if let maxVideoBitrate = abrSettings?.maxVideoBitrate {
            kalturaPlayerRN?.kalturaPlayer?.settings.network.preferredPeakBitRate = maxVideoBitrate
        }
    }
    
    @objc func resetAbrSettings() {
        // The default value in PlayKit for preferredPeakBitRate is 0.
        kalturaPlayerRN?.kalturaPlayer?.settings.network.preferredPeakBitRate = 0
    }
    
    @objc func updateLowLatencyConfig(_ lowLatencyConfig: String?) {
        guard let config = lowLatencyConfig, !config.isEmpty else { return }
        
        let data = Data(config.utf8)
        let pkLowLatencyConfig = try? JSONDecoder().decode(PKLowLatencyConfig.self, from: data)
        if let targetOffsetMs = pkLowLatencyConfig?.targetOffsetMs {
            kalturaPlayerRN?.kalturaPlayer?.settings.lowLatency.targetOffsetMs = targetOffsetMs
        }
    }
    
    @objc func resetLowLatencyConfig() {
        // The default value in PlayKit for targetOffsetMs is 0.
        kalturaPlayerRN?.kalturaPlayer?.settings.lowLatency.targetOffsetMs = 0
    }
}

extension KalturaPlayerModule {
    
    @objc func getCurrentPosition(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            if let currentTime = self?.kalturaPlayerRN?.kalturaPlayer?.currentTime {
                resolve(currentTime)
            } else {
                let error = KalturaPlayerRNError.retrieveCurrentPositionFailed.asNSError
                let message = error.localizedDescription
                reject("ERROR_GETCURRENTPOSITION", message, error)
            }
        }
    }
    
    @objc func isPlaying(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            if let isPlaying = self?.kalturaPlayerRN?.kalturaPlayer?.isPlaying {
                resolve(isPlaying)
            } else {
                let error = KalturaPlayerRNError.retrieveIsPlayingFailed.asNSError
                let message = error.localizedDescription
                reject("ERROR_ISPLAYING", message, error)
            }
        }
    }
    
    @objc func isLive(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            if let isLive = self?.kalturaPlayerRN?.kalturaPlayer?.isLive() {
                resolve(isLive)
            } else {
                let error = KalturaPlayerRNError.retrieveIsLiveFailed.asNSError
                let message = error.localizedDescription
                reject("ERROR_ISPLAYING", message, error)
            }
        }
    }
}

extension KalturaPlayerModule {
    @objc func setLogLevel(logLevel: String?) {
        guard let logLevel = logLevel, !logLevel.isEmpty else {
            return
        }
        
        // TODO:
        
        
    }
}
